### PR TITLE
Multilang asr tutorial

### PR DESCRIPTION
**Pull request by bmwshop**
# What does this PR do ?

Tutorial for inference and training of multilingual ASR models

**Collection**:  ASR

# Changelog 
- Adding the jupyter notebook and a link to it in the readme

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by titu1994 — COMMENTED** [review]
Some comments on the tutorial - 
1) The images for train and infer need to be a bit more aligned - for example the train lang1 tokenizer block is larger than the lang2 tokenizer block
2) When printing configs in the Inference section - use print(OmegaConf.to_yaml(model.cfg.XYZ)) for cleaner output.
3) Dont use single character variable names - m -> model or asr_model. 
4) Dont write "obviously" in documentation, its not obvious to anyone at first glance :P
5) Separate the Tokenizer section into subsections - one for pretrained multilingual model, and another for how to create a multilingual tokenizer out of a pretrained monolingual one.
6) For the tokenizer setup ... I believe the current design works but is a little finicky. Lets look into if we want to make a breaking change for NeMo 1.10 or later and make the change_vocabulary() setup for tokenizer based models more generic.
7) For the section where you show that the tokenizer after changing vocab can no longer transcribe either english or Spanish - give a few words discussion as to why this happens. Primarily because the decoder + joint weights are now larger, and they don't correspond to the original weights. On top of this, the RNNT Blank token is now shared between the two languages, not duplicated at the end of each languages range. Since this is more of an explanation, you can make this section optional for readers who wish to learn about this.
8) For the training section, future proof by updating the Trainer call signatures - gpus -> devices, strategy etc.
9) Remove comments from the train_ds config section, setup_multiple_validation_data() etc.
10) Is computing eval loss necessary? It would be expensive on memory right?
11) Wandb, while much preferred over Tensorboard, is still an optional extension. Check if wandb login was performed or not, and disable wandb logging if its not setup by the user.
12) The WER range should be provided, not a point sample. Given initialiations and different training steps which may occur, give a broader band of expected scores.

**Review by VahidooX — CHANGES_REQUESTED** [review]

**Comment by erastorgueva-nv** [comment]
Hey Dima, looks great. Here are a few minor comments from me, roughly in order of where they appear in the notebook.

1. “Apriori” —> “a priori”. I would put it in italics.

2. “And clearly, the total number of classes the model is trained to output is simply sum of the tokenizer vocabulary lengths.” It needs to be ‘the sum’. Also maybe remove the ‘and clearly’ bit.

3. “multilingual ASR models; e.g. models that can trascribe” —> “i.e.” instead of “e.g.”? “audio” instead of “audios”. typo in “transcribe”.

4. two general ways to creating -> “to create”

5. The diagram illustrates how this works at inference time -> ‘The diagram below’/this diagram'

 6. For instance, if the token id generated by the model is 100, -> ‘For instance, in the example in the diagram, …’

7. I would make lang1 and lang2 italicised or monospaced.

8. !head -3 commonvoice_dev_manifest.json -> This cell should probably be deleted. Or move it up/down and add some markdown explaining what it shows

9. en_files = ["/content/datasets/mini/LibriSpeech/dev-clean-2-processed/7976-110523-0000.wav", "/content/datasets/mini/LibriSpeech/dev-clean-2-processed/7976-110523-0001.wav”]. This doesn’t work on a local machine. I would change it to relative paths (“./datasets/…”). Same for es_files

10. Typo in  “dir: es_tokenizer_dor” —> es_tokenizer_dir 

11.  “The only difference is that you would need to specify the language [id] “ —> No ‘would’, just need. I would get rid of the square brackets and make them rounded instead, so “language (id)”.  Square brackets make me think it is some kind of Python syntax. I would use round brackets here, and everywhere else in the natural language.

12. Remove ‘#’ in # {"audio_filepath": "/data/datasets/LibriSpeech/dev-clean-processed/422-122949-0002.wav", "duration": 4.475, "text": "we truthful ones the nobility in ancient greece called themselves", "lang": "en"}
need to specify the language [id] for each sample in the manifest,

13. “You will noticed later that” —> “Will notice”


14. In the m.change_vocabulary cell, it prints the tokens. Some of them are not English or Spanish characters. I would mention this in the following markdown cell. Something like ‘You may notice that some of the tokens contain characters that are neither part of the English nor Spanish alphabet. This is because the official Mozilla Common Voice data for each language sometimes contains a few utterances from a different language due to some data processing errors. Ideally we would remove these other-language utterances before training. For the purposes of this tutorial, we have skipped this step’.


15. “use them for training and eval”. There is a missing ‘.’


16. “Same needs to done to Spanish manifests.” —> “to be done to the Spanish manifests”

17. “the outcome will be tricky”. This phrase doesn’t make sense to me. Maybe something like “the WER will not be as good”? “Getting the desired WER will be tricky”?


18. “The validation dataset, same thing. The batch size for validation could be higher since memory pressure is lower” —> “For the validation dataset, we follow a similar process. The batch size for validation can be higher…”

19. “not to pollute jupyter output” —> “In order to not pollute the Jupiter notebook output.”
20. In the cell 
```
ENGLISH_TOKENIZER_DIR = "tokenizers/en"
SPANISH_TOKENIZER_DIR = "tokenizers/es"
!mkdir -p ${ENGLISH_TOKENIZER_DIR}
!mkdir -p ${SPANISH_TOKENIZER_DIR}
```
I get an error because it tries to make the dirs '/en', '/es'. There is no error if I make it like `!mkdir -p $ENGLISH_TOKENIZER_DIR`.

21. In the cell 
```
if not os.path.exists("process_asr_text_tokenizer.py"):
  !wget  https://raw.githubusercontent.com/NVIDIA/NeMo/$BRANCH/scripts/tokenizers/process_asr_text_tokenizer.py
```
you get an error if you haven't run the first cell because the variable $BRANCH will be undefined.

**Comment by erastorgueva-nv** [comment]
Also I think the images for the diagrams should be saved as png files in the 'Nemo/tutorials/asr/images' folder.

File: tutorials/asr/README.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ In this repository, you will find several tutorials discussing what is Automatic
 
 11) `Offline_ASR_with_VAD_for_CTC_models`: In this tutorial, we will demonstrate how to use offline VAD to extract speech segments and transcribe the speech segments with CTC models. This will help to exclude some non_speech utterances and could save computation resources by removing unnecessary input to the ASR system.
 
+12) `Multilang_ASR`: We will learn how to work with multilingual ASR models and how to train or fine tune new ones. This noteboks is based on the RNNT Conformer model with BPE Tokenizers.
```
[the remark this answers is not in the document]
Comment (bmwshop) [reply tutorials/asr/README.md]: this is done, thank you.

**Comment by bmwshop** [comment]
Thank you very much @titu1994 , @VahidooX , @erastorgueva-nv , addressed  / incorporated all the proposed changes!

**Review by VahidooX — APPROVED** [review]
LGTM!

**Review by titu1994 — APPROVED** [review]
Looks fantastic !
